src/dune_engine/load_rules.ml

```diff
@@ -688,47 +689,44 @@ end = struct
          | None -> ()
          | Some target_name ->
            report_rule_internal_dir_conflict (Path.Build.basename target_name) loc);
-        match mode with
-        | Standard | Fallback -> acc
-        | Ignore_source_files ->
-          { files = Path.Build.Set.union acc.files targets.files
-          ; dirnames = Filename.Set.union acc.dirnames (Lazy.force target_dirnames)
-          }
-        | Promote { only; _ } ->
-          (* Note that the [only] predicate applies to the files inside the
-             directory targets rather than to directory names themselves. *)
-          let target_files =
-            match only with
-            | None -> targets.files
-            | Some pred ->
-              let is_promoted file =
-                Predicate.test pred (Path.reach (Path.build file) ~from:(Path.build dir))
-              in
-              Path.Build.Set.filter targets.files ~f:is_promoted
-          in
-          { files = Path.Build.Set.union acc.files target_files
-          ; dirnames = Filename.Set.union acc.dirnames (Lazy.force target_dirnames)
-          })
+        (match mode with
+         | Standard | Fallback -> iter ~files ~dirnames rules
+         | Ignore_source_files ->
+           iter
+             ~files:(Path.Build.Set.union files targets.files)
+             ~dirnames:(Filename.Set.union dirnames (Lazy.force target_dirnames))
+             rules
+         | Promote { only; _ } ->
+           (* Note that the [only] predicate applies to the files inside the
+              directory targets rather than to directory names themselves. *)
+           let target_files =
+             match only with
+             | None -> targets.files
+             | Some pred ->
+               let is_promoted file =
+                 Predicate.test pred (Path.reach (Path.build file) ~from:(Path.build dir))
+               in
+               Path.Build.Set.filter targets.files ~f:is_promoted
+           in
+           iter
+             ~files:(Path.Build.Set.union files target_files)
+             ~dirnames:(Filename.Set.union dirnames (Lazy.force target_dirnames))
+             rules)
+    in
+    iter ~files:Path.Build.Set.empty ~dirnames:Filename.Set.empty rules
   ;;
 
   module Source_rules = struct
```

Comment:
This module name could be improved

Reply:
Yeah, might as well do that. We could go even further and unify this type with the `source_paths_to_ignore` type, since they are roughly the same. But I don't think that change belongs in the PR.